File: backend/src/python/adapters/analyst_consensus.ts
/**
 * analyst_consensus adapter — deterministic aggregation of broker
 * recommendations. Ported from
 * python-services/src/financex/calculators/analyst_consensus.py.
 *
 * Needs a list of AnalystReport-shaped rows (broker, recommendation,
 * target_price, report_date) that the current orchestrator does not
 * collect — the LLM agent scrapes is_yatirim + KAP broker filings.
 * Python version consumes
 * accumulatedContext['analyst_consensus_reports'] when operator has
 * pre-seeded it; otherwise emits empty output + warning so the LLM
 * layer can still provide narrative.
 */

export type Recommendation =
  | 'buy' | 'outperform' | 'hold' | 'underperform' | 'sell' | 'unknown';


export interface AnalystReport {
  broker?: string;
  recommendation?: Recommendation | string;
  target_price?: number | string | null;
  report_date?: string;        // ISO YYYY-MM-DD
  horizon_months?: number;
}


export interface LegacyAnalystConsensusOutput {
  agent_id: string;
  output_id: string;
  ticker: string;
  count: number;
  target_price_mean: number | null;
  target_price_median: number | null;
  target_price_high: number | null;
  target_price_low: number | null;
  target_price_stddev: number | null;
  upside_vs_last_close_pct: number | null;
  distribution_buy: number;
  distribution_hold: number;
  distribution_sell: number;
  revision_trend: 'rising' | 'falling' | 'stable' | null;
  crowded_long_flag: boolean;
  reports: AnalystReport[];
  warnings: string[];
  review_status: string;
  source: 'python';
}


const BUY_SET  = new Set<string>(['buy', 'outperform']);
const SELL_SET = new Set<string>(['sell', 'underperform']);


function unwrap<T>(raw: unknown): T | null {
  if (raw == null) return null;
  if (typeof raw === 'string') {
    try { return JSON.parse(raw) as T; } catch { return null; }
  }
  return typeof raw === 'object' ? (raw as T) : null;
}


function parseNumber(v: unknown): number | null {
  if (v == null || v === '') return null;
  const n = typeof v === 'number' ? v : Number(String(v).replace(/[, ]/g, ''));
  return Number.isFinite(n) ? n : null;
}


export function extractReports(raw: unknown): AnalystReport[] {
  const parsed = unwrap<unknown>(raw);
  if (Array.isArray(parsed)) return parsed as AnalystReport[];
  if (parsed && typeof parsed === 'object') {
    const obj = parsed as Record<string, unknown>;
    if (Array.isArray(obj.reports)) return obj.reports as AnalystReport[];
  }
  return [];
}


function bucketCounts(reports: AnalystReport[]): { buy: number; hold: number; sell: number } {
  let buy = 0, hold = 0, sell = 0;
  for (const r of reports) {
    const rec = String(r.recommendation ?? '').toLowerCase();
    if (BUY_SET.has(rec)) buy += 1;
    else if (SELL_SET.has(rec)) sell += 1;
    else if (rec === 'hold') hold += 1;
  }
  return { buy, hold, sell };
}


function round4(n: number): number { return Math.round(n * 10000) / 10000; }


function targetStats(reports: AnalystReport[]): {
  mean: number | null; median: number | null; high: number | null; low: number | null; stddev: number | null;
} {
  const prices = reports.map(r => parseNumber(r.target_price)).filter((v): v is number => v != null);
  if (prices.length === 0) return { mean: null, median: null, high: null, low: null, stddev: null };

  const sorted = [...prices].sort((a, b) => a - b);
  const mean = round4(prices.reduce((a, b) => a + b, 0) / prices.length);
  const median = round4(sorted.length % 2
    ? sorted[(sorted.length - 1) / 2]
    : (sorted[sorted.length / 2 - 1] + sorted[sorted.length / 2]) / 2);
  const high = sorted[sorted.length - 1];
  const low = sorted[0];
  const stddev = prices.length > 1
    ? round4(Math.sqrt(prices.reduce((acc, v) => acc + (v - mean) ** 2, 0) / prices.length))
    : 0;
  return { mean, median, high, low, stddev };
}


function revisionTrend(reports: AnalystReport[], asOfIso?: string): 'rising' | 'falling' | 'stable' | null {
  const priced = reports.filter(r => parseNumber(r.target_price) != null && r.report_date);
  if (priced.length < 3) return null;

  const parseDate = (iso: string) => new Date(iso).getTime();
  const refMs = asOfIso
    ? parseDate(asOfIso)
    : Math.max(...priced.map(r => parseDate(r.report_date!)));
  if (!Number.isFinite(refMs)) return null;

  const DAY = 86_400_000;
  const recentCutoff = refMs - 30 * DAY;
  const olderCutoff  = refMs - 90 * DAY;

  const recent: number[] = [];
  const older:  number[] = [];
  for (const r of priced) {
    const t = parseDate(r.report_date!);
    const price = parseNumber(r.target_price)!;
    if (t >= recentCutoff) recent.push(price);
    else if (t >= olderCutoff) older.push(price);
  }
  if (recent.length < 2 || older.length < 2) return null;

  const meanOf = (xs: number[]) => xs.reduce((a, b) => a + b, 0) / xs.length;
  const recentMean = meanOf(recent);
  const olderMean  = meanOf(older);
  if (olderMean === 0) return null;
  const delta = (recentMean - olderMean) / Math.abs(olderMean);
  if (delta > 0.03)  return 'rising';
  if (delta < -0.03) return 'falling';
  return 'stable';
}


export function adaptAnalystConsensusForLegacy(
  reports: AnalystReport[],
  ticker: string,
  outputId: string,
  opts: { lastClose?: number | null; asOf?: string | null } = {},
): LegacyAnalystConsensusOutput {
  const warnings: string[] = [];
  if (reports.length === 0) {
    warnings.push('No analyst reports supplied — accumulatedContext[analyst_consensus_reports] empty. LLM layer should run broker scrape.');
  }

  const dist = bucketCounts(reports);
  const stats = targetStats(reports);
  const trend = revisionTrend(reports, opts.asOf ?? undefined);

  let upside: number | null = null;
  const lc = opts.lastClose;
  if (lc != null && lc > 0 && stats.mean != null) {
    upside = Math.round(((stats.mean / lc) - 1) * 100 * 100) / 100;
  }

  // Crowded long: 5+ reports and zero SELLs — same heuristic as
  // sentiment_news crowded-long flag.
  const crowdedLong = reports.length >= 5 && dist.sell === 0 && dist.buy >= 3;
  if (crowdedLong) {
    warnings.push(`Crowded long — zero SELL ratings across ${reports.length} analysts`);
  }

  return {
    agent_id: 'analyst_consensus_agent',
    output_id: outputId,
    ticker: ticker.toUpperCase(),
    count: reports.length,
    target_price_mean: stats.mean,
    target_price_median: stats.median,
    target_price_high: stats.high,
    target_price_low: stats.low,
    target_price_stddev: stats.stddev,
    upside_vs_last_close_pct: upside,
    distribution_buy: dist.buy,
    distribution_hold: dist.hold,
    distribution_sell: dist.sell,
    revision_trend: trend,
    crowded_long_flag: crowdedLong,
    reports,
    warnings,
    review_status: 'pending_ceo_review',
    source: 'python',
  };
}

File: python-services/src/financex/calculators/analyst_consensus.py
"""analyst_consensus — deterministic aggregation of broker reports.

Python takes a list of AnalystReport rows (from KAP broker filings,
is_yatirim research page, or a manual JSON fixture) and produces
canonical consensus statistics + a small set of red-flag signals.
The LLM layer later writes the narrative — *why* specific brokers
recommend SELL, what a 30% target spread implies, etc.

Key signals the LLM should see:
  - `distribution_*` counts let it call 'crowded long' when SELLs = 0.
  - `target_price_stddev` flags unusual disagreement across the Street.
  - `revision_trend` ('rising' / 'falling' / 'stable') is derived from
    report dates when ≥2 cohorts are present.
"""

from __future__ import annotations

import statistics
from datetime import date, timedelta
from decimal import Decimal

from financex.schemas.analyst import AnalystConsensus, AnalystReport, Recommendation


# ---------------------------------------------------------------------
# Distribution buckets
# ---------------------------------------------------------------------

_BUY_SET = {Recommendation.BUY, Recommendation.OUTPERFORM}
_SELL_SET = {Recommendation.SELL, Recommendation.UNDERPERFORM}


def _bucket_counts(reports: list[AnalystReport]) -> tuple[int, int, int]:
    buy = sum(1 for r in reports if r.recommendation in _BUY_SET)
    sell = sum(1 for r in reports if r.recommendation in _SELL_SET)
    hold = sum(1 for r in reports if r.recommendation == Recommendation.HOLD)
    return buy, hold, sell


# ---------------------------------------------------------------------
# Target-price statistics
# ---------------------------------------------------------------------

def _target_stats(
    reports: list[AnalystReport],
) -> tuple[Decimal | None, Decimal | None, Decimal | None, Decimal | None, Decimal | None]:
    """Return (mean, median, high, low, stddev) of non-null target prices."""
    prices = [r.target_price for r in reports if r.target_price is not None]
    if not prices:
        return (None, None, None, None, None)
    mean = Decimal(str(round(statistics.fmean(float(p) for p in prices), 4)))
    median = Decimal(str(round(statistics.median(float(p) for p in prices), 4)))
    high = max(prices)
    low = min(prices)
    stddev = (
        Decimal(str(round(statistics.pstdev(float(p) for p in prices), 4)))
        if len(prices) > 1
        else Decimal("0")
    )
    return mean, median, high, low, stddev


# ---------------------------------------------------------------------
# Revision trend — compare recent 30 days vs preceding 60
# ---------------------------------------------------------------------

def _revision_trend(reports: list[AnalystReport], *, as_of: date | None = None) -> str | None:
    """Return 'rising' / 'falling' / 'stable' / None."""
    pricy = [r for r in reports if r.target_price is not None]
    if len(pricy) < 3:
        return None

    ref = as_of or max(r.report_date for r in pricy)
    recent_cutoff = ref - timedelta(days=30)
    older_cutoff = ref - timedelta(days=90)

    recent = [r.target_price for r in pricy if r.report_date >= recent_cutoff]
    older = [
        r.target_price
        for r in pricy
        if older_cutoff <= r.report_date < recent_cutoff
    ]
    if len(recent) < 2 or len(older) < 2:
        return None

    recent_mean = statistics.fmean(float(p) for p in recent if p is not None)
    older_mean = statistics.fmean(float(p) for p in older if p is not None)
    if older_mean == 0:
        return None
    delta_pct = (recent_mean - older_mean) / abs(older_mean)
    if delta_pct > 0.03:
        return "rising"
    if delta_pct < -0.03:
        return "falling"
    return "stable"


# ---------------------------------------------------------------------
# Public entry point
# ---------------------------------------------------------------------

def aggregate_reports(
    reports: list[AnalystReport],
    *,
    last_close: Decimal | None = None,
    as_of: date | None = None,
) -> AnalystConsensus:
    """Build an AnalystConsensus over a list of broker notes.

    `last_close` (optional) enables `upside_vs_last_close_pct` on the
    consensus — a hand-off the LLM uses when writing the "target
    upside" line in the report.
    """
    buy, hold, sell = _bucket_counts(reports)
    mean, median, high, low, stddev = _target_stats(reports)
    trend = _revision_trend(reports, as_of=as_of)

    upside: Decimal | None = None
    if last_close is not None and last_close > 0 and mean is not None:
        upside = (((mean / last_close) - Decimal("1")) * Decimal("100")).quantize(Decimal("0.01"))

    return AnalystConsensus(
        reports=list(reports),
        count=len(reports),
        target_price_mean=mean,
        target_price_median=median,
        target_price_high=high,
        target_price_low=low,
        target_price_stddev=stddev,
        upside_vs_last_close_pct=upside,
        distribution_buy=buy,
        distribution_hold=hold,
        distribution_sell=sell,
        revision_trend=trend,
    )
